Output intepretation 
---------------------

``Result`` folder under current path(default) or output\_folder set by
user. A typical structure of ``Result`` is follows:

::

    Result
    ├── Alignment
    │   ├── Bowtie2
    │   │   └── bowtie2_unmapped_S483_T_chrX_R.bam
    │   ├── BWA
    │   │   └── bwa_S483_T_chrX_R.mem.sam
    │   └── STAR
    │       └── star_S483_T_chrX_R_Chimeric.out.junction
    ├── circRNA_Identification
    │   ├── CIRCexplorer2
    │   │   ├── circexplorer2_merge.matrix
    │   │   ├── CIRCexplorer2_S483_T_chrX_R_circularRNA_known.txt
    │   │   └── S483_T_chrX_R_modify_circexplorer2.candidates.bed
    │   ├── CIRI
    │   │   ├── ciri_merge.matrix
    │   │   ├── CIRI_S483_T_chrX_R.txt
    │   │   └── S483_T_chrX_R_modify_ciri.candidates.bed
    │   ├── Find_circ
    │   │   ├── find_circ_merge.matrix
    │   │   ├── find_circ_S483_T_chrX_R_splice_sites.bed
    │   │   └── S483_T_chrX_R_modify_finc_circ.candidates.bed
    │   ├── Mapsplice
    │   │   ├── mapsplice_merge.matrix
    │   │   ├── output_mapsplice_S483_T_chrX_R
    │   │   │   ├── alignments.sam
    │   │   │   ├── circular_RNAs.txt
    │   │   │   ├── deletions.txt
    │   │   │   ├── fusions_candidates.txt
    │   │   │   ├── fusions_not_well_annotated.txt
    │   │   │   ├── fusions_raw.txt
    │   │   │   ├── fusions_well_annotated.txt
    │   │   │   ├── insertions.txt
    │   │   │   ├── junctions.txt
    │   │   │   ├── logs
    │   │   │   │   ├── add_fusion_strand_consistent.err
    │   │   │   │   ├── add_fusion_strand_consistent.log
    │   │   │   │   ├── alignment_handler_fusion.err
    │   │   │   │   ├── alignment_handler_fusion.log
    │   │   │   │   ├── alignment_handler_remap.err
    │   │   │   │   ├── alignment_handler_remap.log
    │   │   │   │   ├── anchor_bowtie.err
    │   │   │   │   ├── anchor_bowtie.log
    │   │   │   │   ├── annotate_fusion.log
    │   │   │   │   ├── best_junction_semi_non_canon_remained_ROC.log
    │   │   │   │   ├── bowtie_build.log
    │   │   │   │   ├── bowtie_refnames
    │   │   │   │   ├── check_reads_format.log
    │   │   │   │   ├── collectstats.log
    │   │   │   │   ├── FilterFusionByNormalPaired.log
    │   │   │   │   ├── FilterFusionSamByFusionJunc.log
    │   │   │   │   ├── filterremappedfusion.log
    │   │   │   │   ├── fusion_junc_db.log
    │   │   │   │   ├── fusionsam2junc_original_fusion_junction.log
    │   │   │   │   ├── fusionsam2junc_remapped_fusion_junction.log
    │   │   │   │   ├── fusion_unmapped.1_2sam.log
    │   │   │   │   ├── fusion_unmapped.2_2sam.log
    │   │   │   │   ├── gtf2genetab.log
    │   │   │   │   ├── junc_db.log
    │   │   │   │   ├── load_fusion_chrom_seq_std.log
    │   │   │   │   ├── load_fusion_chrom_seq_std_long_seq.log
    │   │   │   │   ├── long_fusion_seq_bowtie.err
    │   │   │   │   ├── long_fusion_seq_bowtie.log
    │   │   │   │   ├── mapsplice_fusion.log
    │   │   │   │   ├── mapsplice_fusion_original.log
    │   │   │   │   ├── mapsplice_original.log
    │   │   │   │   ├── mapsplice_remap.log
    │   │   │   │   ├── matchfusion2normal.log
    │   │   │   │   ├── ori.all_junctions_filtered_by_min_mis_lpq.log
    │   │   │   │   ├── ori.filteroriginalfusion.log
    │   │   │   │   ├── pair_repeat_reads.err
    │   │   │   │   ├── pair_repeat_reads.log
    │   │   │   │   ├── read_chromo_sizes.log
    │   │   │   │   ├── remap_cluster.log
    │   │   │   │   ├── remap_parseCluster.log
    │   │   │   │   ├── sam2junc_ori.all_junctions.log
    │   │   │   │   ├── SeparateNormalFromFusionJunc.log
    │   │   │   │   ├── SetUnmappedBitFlag.log
    │   │   │   │   └── sort_repeat.log
    │   │   │   └── stats.txt
    │   │   ├── S483_T_chrX_R_mapsplice.log
    │   │   └── S483_T_chrX_R_modify_mapsplice.candidates.bed
    │   └── Segemehl
    │       ├── S483_T_chrX_R_modify_segemehl.candidates.bed
    │       ├── segemehl_merge.matrix
    │       └── segemehl_S483_T_chrX_R_splicesites.bed
    ├── Combination_Matrix
    │   ├── all_tools_merge.matrix
    │   └── final.matrix
    ├── Annotation
    │   ├── merge_for_annotation_annote.txt
    │   ├── merge_venn.png  
    │   ├── merge_distribution.png 
    │   ├── merge_spanningtree.png 
    │   ├── merge_hist.png 
    │   ├── merge_circos.png 
    │   └── merge_calculates.pdf
    ├── DE_Analysis
    │   ├── merge_volcano.png 
    │   ├── merge_heatmap1.png 
    │   ├── merge_heatmap2.png 
    │   ├── merge_heatmap3.png 
    │   ├── merge_pca1.png 
    │   ├── merge_pca2.png
    │   └── merge_plots.pdf
    ├── Correlation_Analysis
    └── QC
        ├── multiqc_report.html
        └── S483_T_chrX_R_fastpreport.html

-  ``QC`` stored the Quality control output generated by Fastp software.
-  ``circRNA_Identification`` contains all assembled circRNAs and their
   sequences. **.candidates.bed\* is the standard bed6 format of the
   known circRNAs; **.matrix\* is the matrix of all samples gene
   expression.
-  ``Alignment`` are BWA/Bowtie2/STAR aligner outputs.
-  ``Combination_Matrix`` stores the matrix merge from all the selected
   tools. *final.matrix* stored the expression matrix and
   *all\_tools\_merge.matrix* show whether the tools are detected in
   this gene.
-  ``Annoation`` stores the information of the circRNAs annotation.
-  ``DE_Analysis`` stores the information of the circRNAs differential
   expression.
-  ``Correlation_Analysis`` stores the information of the circRNAs
   correlation analysis.

**Output directory: ``Result/QC `` ** 

-  ``S483_T_chrX_R_fastpreport.html``
-  Fastp report of raw fastq files
-  ``multiqc_report.html``
-  merge of all Fastp report

**Output directory: ``Result/Combination_Matrix`` .**

-  ``final.matrix``
-  the expression matrix of all samples
-  ``all_tools_merge.matrix``
-  the information of whether the tools are detected in this gene

**Output directory: ``Result/Alignment``**

-  ``*.bam``
-  alignment result in bam format.
-  ``*.sam``
-  alignment result in sam format.

**Output directory: ``Result/circRNA_Identification``**

-  ``*.candidates.bed``
-  the standard bed6 format of the known circRNAs.

-  ``*.matrix``
-  the matrix of all samples gene expression of the specific pipelines.

**Output directory: ``Result/Annoation``**

-  the information of the circRNAs annotation.

**Output directory: ``Result/DE_Analysis``**

-  the information of the circRNAs differential expression.

**Output directory: ``Result/Correlation_Analysis``**

-  the information of the circRNAs correlation analysis.
